code/game/objects/items/weapons/tools/_tools.dm

```diff
@@ -44,11 +97,244 @@
+		return TOOL_USE_CANCEL
+
+
+
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
+		if(H.shock_stage >= 30)
+			user << SPAN_WARNING("Pain distracts you from your task.")
+			fail_chance += round(H.shock_stage/120 * 40)
+			base_time += round(H.shock_stage/120 * 40)
+
+
+	//Start time and time spent are used to calculate resource use
+	var/start_time = world.time
+	var/time_spent = 0
+
+	//Precalculate worktime here
+	var/time_to_finish = 0
+	if (base_time)
+		time_to_finish = base_time - get_tool_quality(required_quality) - user.stats.getStat(required_stat)
+
+	if((instant_finish_tier < get_tool_quality(required_quality)) || time_to_finish < 0)
+		time_to_finish = 0
+
+	var/obj/item/weapon/tool/T
+	if(istype(src, /obj/item/weapon/tool))
+		T = src
+		if(!T.check_tool_effects(user, time_to_finish))
+			return TOOL_USE_CANCEL
+
+	//Repeating sound code!
+	//A datum/repeating_sound is a little object we can use to make a sound repeat a few times
+	var/datum/repeating_sound/toolsound = null
+	if(forced_sound != NO_WORKSOUND)
+
+		var/soundfile
+		if(forced_sound)
+			soundfile = forced_sound
+		else
+			soundfile = worksound
+
+		if (sound_repeat && time_to_finish)
+			//It will repeat roughly every 2.5 seconds until our tool finishes
+			toolsound = new/datum/repeating_sound(sound_repeat,time_to_finish,0.15, src, soundfile, 70, 1)
+		else
+			playsound(src.loc, soundfile, 70, 1)
+
+	//The we handle the doafter for the tool usage
+	if(time_to_finish)
+		target.used_now = TRUE
+
+		if(!do_after(user, time_to_finish, user))
+			//If the doafter fails
+			user << SPAN_WARNING("You need to stand still to finish the task properly!")
+			target.used_now = FALSE
+			time_spent = world.time - start_time //We failed, spent only part of the time working
+			if (T)
+				T.consume_resources(time_spent, user)
+			if (toolsound)
+				//We abort the repeating sound.
+				//Stop function will delete itself too
+				toolsound.stop()
+				toolsound = null
+			return TOOL_USE_CANCEL
+		else
+			target.used_now = FALSE
+
+	//If we get here the operation finished correctly, we spent the full time working
+	time_spent = time_to_finish
+	if (T)
+		T.consume_resources(time_spent, user)
+
+	//Safe cleanup
+	if (toolsound)
+		toolsound.stop()
+		toolsound = null
+
+	var/stat_modifer = 0
+	if(required_stat)
+		stat_modifer = user.stats.getStat(required_stat)
+	fail_chance = fail_chance - get_tool_quality(required_quality) - stat_modifer
+	if(prob(fail_chance))
+		user << SPAN_WARNING("You failed to finish your task with [src.name]! There was a [fail_chance]% chance to screw this up.")
+		return TOOL_USE_FAIL
+
+	return TOOL_USE_SUCCESS
+
+
+
+
+/******************************
+	/* Tool Failure */
+*******************************/
+
+//Critical failure rolls. If you use use_tool_extended, you might want to call that proc as well.
+/obj/item/proc/handle_failure(var/mob/living/user, var/atom/target, required_stat = null, required_quality)
```

Comment:
`required_stat = null` there is no need to assign null in arguments, they are null by default.